lib/include/iot_linear_containers.h

```diff
@@ -0,0 +1,888 @@
+ *
+ * @param[in] pLink The element to remove.
+ */
+/* @[declare_linear_containers_list_double_remove] */
+static inline void IotListDouble_Remove( IotLink_t * const pLink )
+/* @[declare_linear_containers_list_double_remove] */
+{
+    /* This function must not be called with a NULL parameter. */
+    IotContainers_Assert( pLink != NULL );
+
+    /* This function must be called on a linked element. */
+    IotContainers_Assert( IotLink_IsLinked( pLink ) == true );
+
+    pLink->pPrevious->pNext = pLink->pNext;
+    pLink->pNext->pPrevious = pLink->pPrevious;
+    pLink->pPrevious = NULL;
+    pLink->pNext = NULL;
+}
+
+/**
+ * @brief Remove the element at the head of a doubly-linked list.
+ *
+ * @param[in] pList The doubly-linked list that holds the element to remove.
+ *
+ * @return Pointer to an #IotLink_t representing the removed list head; `NULL`
+ * if the list is empty. The macro #IotLink_Container may be used to determine
+ * the address of the link's container.
+ */
+/* @[declare_linear_containers_list_double_removehead] */
+static inline IotLink_t * IotListDouble_RemoveHead( IotListDouble_t * const pList )
+/* @[declare_linear_containers_list_double_removehead] */
+{
+    IotLink_t * pHead = NULL;
+
+    if( IotListDouble_IsEmpty( pList ) == false )
+    {
+        pHead = pList->pNext;
+        IotListDouble_Remove( pHead );
+    }
+
+    return pHead;
+}
+
+/**
+ * @brief Remove the element at the tail of a doubly-linked list.
+ *
+ * @param[in] pList The doubly-linked list that holds the element to remove.
+ *
+ * @return Pointer to an #IotLink_t representing the removed list tail; `NULL`
+ * if the list is empty. The macro #IotLink_Container may be used to determine
+ * the address of the link's container.
+ */
+/* @[declare_linear_containers_list_double_removetail] */
+static inline IotLink_t * IotListDouble_RemoveTail( IotListDouble_t * const pList )
+/* @[declare_linear_containers_list_double_removetail] */
+{
+    IotLink_t * pTail = NULL;
+
+    if( IotListDouble_IsEmpty( pList ) == false )
+    {
+        pTail = pList->pPrevious;
+        IotListDouble_Remove( pTail );
+    }
+
+    return pTail;
+}
+
+/**
+ * @brief Remove all elements in a doubly-linked list.
+ *
+ * @param[in] pList The list to empty.
+ * @param[in] freeElement A function to free memory used by each removed list
+ * element. Optional; pass `NULL` to ignore.
+ * @param[in] linkOffset Offset in bytes of a link member in its container, used
+ * to calculate the pointer to pass to `freeElement`. This value should be calculated
+ * with the C `offsetof` macro. This parameter is ignored if `freeElement` is `NULL`
+ * or its value is `0`.
+ */
+/* @[declare_linear_containers_list_double_removeall] */
+static inline void IotListDouble_RemoveAll( IotListDouble_t * const pList,
+                                            void ( *freeElement )( void * ),
+                                            size_t linkOffset )
```

Comment:
No need to pass offset as a parameter, it can be more efficiently computed with 
IotLink_Container